File: torch_tpu/common/discovery_test.cc
/*
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "torch_tpu/common/discovery.h"

#include <cstdlib>
#include <string>

#include "gtest/gtest.h"
#include "absl/log/absl_check.h"
#include "absl/status/statusor.h"
#include "torch_tpu/common/env_vars.h"
#include "torch_tpu/common/error_utils.h"

namespace torch_tpu {
namespace {

using testing::ExitedWithCode;

// Since GetPremappedBufferSizeFromEnvOnce() memoizes the environment variable,
// we need to use EXPECT_EXIT() to test the behavior of the function when
// setting the environment variable to different values (EXPECT_EXIT() runs
// the function in a subprocess, so side effects are not shared between tests).

TEST(GetPremappedBufferSizeFromEnvOnceDeathTest, Returns0ByDefault) {
  EXPECT_EXIT(
      {
        unsetenv(kTpuPremappedBufferSizeEnvVar);
        TT_ASSIGN_OR_CRASH(auto size, GetPremappedBufferSizeFromEnvOnce());
        exit(size);
      },
      // Verifies that *status_or_size is 0.
      ExitedWithCode(0), "");
}

TEST(GetPremappedBufferSizeFromEnvOnceDeathTest, SetsEnvVarIfNotFound) {
  EXPECT_EXIT(
      {
        unsetenv(kTpuPremappedBufferSizeEnvVar);
        const auto& status_or_size = GetPremappedBufferSizeFromEnvOnce();
        ABSL_CHECK_OK(status_or_size.status());
        // This should be true.
        const bool env_var_is_0 =
            std::string(getenv(kTpuPremappedBufferSizeEnvVar)) == "0";
        exit(env_var_is_0 ? 0 : 1);
      },
      // Verifies that env_var_is_0 is true.
      ExitedWithCode(0), "");
}

TEST(GetPremappedBufferSizeFromEnvOnceDeathTest, ReturnsEnvVarValueIfSet) {
  EXPECT_EXIT(
      {
        setenv(kTpuPremappedBufferSizeEnvVar, "123", 1);
        const auto& status_or_size = GetPremappedBufferSizeFromEnvOnce();
        if (!status_or_size.ok()) {
          exit(1);  // Failed to getenv.
        }
        exit(*status_or_size);
      },
      // Verifies that *status_or_size is 123.
      ExitedWithCode(123), "");
}

}  // namespace
}  // namespace torch_tpu

File: examples/benchmarks/e2e/harness/tests/discovery_test.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for harness/discovery.py."""

import importlib
import pathlib
import sys
import tempfile
import textwrap

from absl.testing import absltest
from examples.benchmarks.e2e.harness import discovery as discovery_lib
from examples.benchmarks.e2e.harness import registry as registry_lib


class DiscoveryTest(absltest.TestCase):

  def _make_pkg(self, modules: dict[str, str]):
    """Makes a throwaway python package for testing import side-effects."""
    tmp = tempfile.mkdtemp()
    pkg_name = f"fake_models_{pathlib.Path(tmp).name}"
    pkg_dir = pathlib.Path(tmp) / pkg_name
    pkg_dir.mkdir()
    (pkg_dir / "__init__.py").write_text("")
    for mod_name, body in modules.items():
      (pkg_dir / f"{mod_name}.py").write_text(textwrap.dedent(body))
    sys.path.insert(0, tmp)
    self.addCleanup(lambda: sys.path.remove(tmp))

    pkg = importlib.import_module(pkg_name)
    self.addCleanup(
        lambda: [
            sys.modules.pop(m, None)
            for m in list(sys.modules)
            if m.startswith(pkg_name)
        ]
    )
    return pkg

  def test_imports_all_submodules(self):
    pkg = self._make_pkg({"a": "LOADED='a'", "b": "LOADED='b'"})
    self.assertEqual(discovery_lib.import_submodules(pkg), [])
    self.assertIn(f"{pkg.__name__}.a", sys.modules)
    self.assertIn(f"{pkg.__name__}.b", sys.modules)

  def test_broken_module_surfaced_not_swallowed(self):
    pkg = self._make_pkg(
        {"good": "LOADED=True", "broken": "raise RuntimeError('boom')"}
    )
    failures = discovery_lib.import_submodules(pkg)
    self.assertLen(failures, 1)
    self.assertIsInstance(failures[0], discovery_lib.ImportFailure)
    self.assertIn("broken", failures[0].module)
    self.assertIn("boom", str(failures[0]))

  def test_one_broken_module_does_not_block_others(self):
    pkg = self._make_pkg(
        {"broken": "raise RuntimeError('boom')", "good": "LOADED=True"}
    )
    discovery_lib.import_submodules(pkg)
    self.assertIn(f"{pkg.__name__}.good", sys.modules)

  def test_registration_happens_via_import(self):
    saved = dict(registry_lib.REGISTRY)
    registry_lib.REGISTRY.clear()
    try:
      pkg = self._make_pkg({
          "m": (
              """
              from examples.benchmarks.e2e.harness import registry as registry_lib
              from examples.benchmarks.e2e.harness import step_lib

              @registry_lib.register_benchmark(
                  stepper=step_lib.StepperType.FORWARD
              )
              def discovered_probe():
                return ("m", (), {}, None)
              """
          ),
      })
      self.assertEqual(discovery_lib.import_submodules(pkg), [])
      self.assertIn("discovered_probe", registry_lib.REGISTRY)
    finally:
      registry_lib.REGISTRY.clear()
      registry_lib.REGISTRY.update(saved)


if __name__ == "__main__":
  absltest.main()
